crates/ra_hir/src/ty.rs

```diff
@@ -499,86 +470,103 @@ pub(super) fn type_for_field(db: &impl HirDatabase, def_id: DefId, field: Name)
 /// The result of type inference: A mapping from expressions and patterns to types.
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct InferenceResult {
-    type_of: FxHashMap<LocalSyntaxPtr, Ty>,
+    type_of_expr: FxHashMap<ExprId, Ty>,
+    type_of_pat: FxHashMap<PatId, Ty>,
```

Comment:
Ids are assigned sequentially, so we could use a `Vec` instead of map here potentially. Might be a good idea to add `ArenaMap` to arena crate, or use `index_map` (or `vec_map` -- there were a crate for this.).